### ISSUE TITLE
ocr of rotated image

### ISSUE THREAD

**Issue by josef821** [robertknight/ocrs#121]
hi, thanks for your useful ocr engine,
its works good but when i try to set rotated image it return bad result.
![Screenshot 2024-09-30 161933](https://github.com/user-attachments/assets/a616df33-2fb7-443a-834d-8e6f5c232f06)

is there any fix tips?

**Comment by robertknight** [comment]
Currently the recognition model and layout logic assumes that the image is approximately upright (some amount of rotation or skew is OK) and that the text is read left to right. To work with rotated or severely skewed images, they need to be rotated / de-skewed as a preprocessing step. Eventually this should be integrated into this library, but in the meantime you could try something like:

1. Call the `OcrEngine::detect_words` method to detect bounding boxes of connected areas (the white regions in the top-left image)
2. Infer the orientation from the positions and aspect ratios of the boxes (eg. if most boxes are tall rather than wide, that means the text is probably upside-down)
3. Use functions in the `imageproc` crate to rotate the image based on the inferred orientation
4. Perform OCR or the rotated image

A more sophisticated approach would be to use an image classification model to infer the orientation of each word, or a sample of words. If a suitable model was created in eg. PyTorch and exported to ONNX, it could then be converted to [RTen](https://github.com/robertknight/rten) and used in the above preprocessing pipeline instead of heuristics.

**Comment by josef821** [comment]
thanks for reply.
i will do that. i check all masks to check lines are rotated or not.
your layout analyze is not good enough. i check you reply to other. you want to create a model for cluster and sort word to get line bounding box. How long do you think it will take to be able to publish the layout analyze model with its training code?

**Comment by robertknight** [comment]
>  How long do you think it will take to be able to publish the layout analyze model with its training code?

I don't know. All the code that exists is in the [ocrs-models](https://github.com/robertknight/ocrs-models) repository, but for layout analysis that only includes some non-functional prototypes.

In the meantime, if you happen to be working with documents that have a predictable layout, you can always substitute the [`find_text_lines`](https://docs.rs/ocrs/latest/ocrs/struct.OcrEngine.html#method.find_text_lines) step with custom code.

**Comment by josef821** [comment]
exist layout analysis not working good for curve layout or complex image. i waiting for your layout analysis.
thanks
